### PR TITLE
Organelle Finder workflow

### PR DESCRIPTION
**Pull request by ViktorSade**
Prototype organelle finder workflow

### CODE REVIEW

File: subworkflows/OrganelleFinder/OrganelleFinder.nf [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+params.genome_assembly = ''
+params.outdir = 'results'
+
+// tblastn parameters
+params.reference_organelle = ''
+params.blast_evalue = '1e-21'
+
+// filter parameters
+params.bitscore = 150
+params.significant_gene_matches = 3
```
Comment (mahesh-panchal) [inline subworkflows/OrganelleFinder/OrganelleFinder.nf:23]: Please put these in the `nextflow.config` file.

File: subworkflows/OrganelleFinder/OrganelleFinder.nf [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+            .ifEmpty { exit 1, "Cannot find genome matching ${params.genome_assembly}!\n" }
+            .set {genome_assembly}
+        Channel.fromPath(params.reference_organelle, checkIfExists: true)
+            .ifEmpty { exit 1, "Cannot find blast database files matching ${params.reference_organelle}!\n" }
+            .set {reference_organelle}
+        organelle_finder(genome_assembly,reference_organelle)
+
+}
+
+workflow organelle_finder {
```
Comment (mahesh-panchal) [inline subworkflows/OrganelleFinder/OrganelleFinder.nf:59]: Please use uppercase for workflow and process names.
```suggestion
workflow ORGANELLE_FINDER {
```

**Review by mahesh-panchal — COMMENTED** [review]
Just some general comments for now.

- Use a `.gitignore` file to make sure log files, data, etc are not included. Only workflow code and configuration should be part of the pull request. 

- Please use the format for processes that is in the HackMD document. ( Uppercase for process names, container and conda in the process file, etc). It's a change from what our current system is, but I will also make that change in future to the rest of the workflows. 
- Put each process in it's own file in a modules directory. 
- Use named output channels, e.g. `blast_result = EXTRACT.out.blast_result`

**Review by mahesh-panchal — COMMENTED** [review]
- Use Module aliasing to reduce duplication. 
    ```nextflow
    include { TBLASTN as TBLASTN_CHLOROPLAST; TBLASTN as TBLASTN_MITOCHONDRIA }
    ```
- Perhaps add an additional parameter to the processes to determine what to run.e.g:
    ```nextflow
    TBLAST( ... , 'chloroplast' )
 
   ...
    
    process TBLASTN {
        input:
        ...
        val organelle  // possible values: 'chloroplast','mitochondria'
        ...
    }
    ```
- Remove the files not being used like the`software_packages.config`.
- pass all `params.<parameter>` as process inputs (i.e. only reference params in a workflow block), and do not access them directly in the scripts (e.g.params.bitscore).
- The code block pattern:
    ```
    LINES=\$(cat $accessions_matchfiltered)
    for line in \$LINES
    do
        grep \$line $statistics_bitfiltered >> chloroplast_significant_matches_statistics.tsv
    done
    ```
    can be replaced by:
    ```
    grep -f <pattern_file> <file_to_search>
    ```

File: subworkflows/OrganelleFinder/nextflow.config [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+}
+
+params {
+    // Input genome paths
+    params.genome_assembly = ''
+    params.reference_mitochondria = ''
+    params.reference_chloroplast = ''
+
+    // Output directory
+    params.outdir = ''
```
Comment (mahesh-panchal) [inline subworkflows/OrganelleFinder/nextflow.config:18]:
```suggestion
    genome_assembly = ''
    reference_mitochondria = ''
    reference_chloroplast = ''

    // Output directory
    outdir = ''
```

**Review by mahesh-panchal — COMMENTED** [review]
Good work. 
1. Remove the non-portable files ( *.yml ) from the repo.
2. Please update the README with usage instructions and a description of what it does.
If you have time
3. Use more whitespace to help readabilty. Some guidelines here: https://github.com/carpentries-incubator/workflows-nextflow/blob/9e269d438a65bbff1b312f91431d154ffd3b9f9e/_episodes/15-coding_practices.md
4. Move publishing criteria to the config.

File: subworkflows/OrganelleFinder/modules/extract_final.nf [hunk 4]
```diff
@@ -0,0 +1,39 @@
+process EXTRACT_FINAL {
+
+    publishDir "${outdir}", mode: 'copy', pattern: "${assembly.baseName}_${organelle}.fna"
+    publishDir "${outdir}", mode: 'copy', pattern: "${assembly.baseName}_nuclear.fna"
+
+    input:
+    path assembly
+    path matched_accessions
+    path outdir
```
Comment (mahesh-panchal) [inline subworkflows/OrganelleFinder/modules/extract_final.nf:9]: This is an unnecessary parameter. Moving `publishDir` to a config should clean this up.
The config should look like this
```groovy
process {
    withName: 'EXTRACT_FINAL' {
        publishDir = [
            [ 
                path: params.outdir,
                mode: 'copy',
                pattern: { "${assembly.baseName}_${organelle}.fna" }
            ],
            [
                path: params.outdir,
                mode: 'copy',
                pattern: { "${assembly.baseName}_nuclear.fna" }
            ]
        ]
    }
}
```

File: subworkflows/OrganelleFinder/modules/filter.nf [hunk 5]
```diff
@@ -0,0 +1,50 @@
+process FILTER {
+    
+    publishDir "${outdir}/statistics", mode: 'copy', pattern: "${organelle}_statistics_summary.tsv"
+
```
Comment (mahesh-panchal) [inline subworkflows/OrganelleFinder/modules/filter.nf:4]: Add a default container to these processes.
```groovy
    conda (params.enable_conda ? "conda-forge::sed=4.7" : null)
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://containers.biocontainers.pro/s3/SingImgsRepo/biocontainers/v1.2.0_cv1/biocontainers_v1.2.0_cv1.img' :
        'biocontainers/biocontainers:v1.2.0_cv1' }"
```

File: subworkflows/OrganelleFinder/modules/tblastn.nf [hunk 6]
```diff
@@ -0,0 +1,25 @@
+process TBLASTN {
+
+    conda "${task.ext.enable_conda ? 'bioconda::tool=blast:2.12.0' : '' }"
```
Comment (mahesh-panchal) [inline subworkflows/OrganelleFinder/modules/tblastn.nf:3]:
```suggestion
    conda "${task.ext.enable_conda ? 'bioconda::blast:2.12.0' : '' }"
```
Please check it works under conda too.

File: subworkflows/OrganelleFinder/modules/mapping.nf [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+
+    output:
+    path "depth_full_alignment.tsv", emit: depth_file
+
+    script:
+    """
+    minimap2 -d assembly.mmi $assembly
+    minimap2 -a assembly.mmi $reads_file > assembly_full_alignment.sam
+    samtools view -S -b assembly_full_alignment.sam > assembly_full_alignment.bam
+    samtools sort -o sorted_assembly_full_alignment.bam assembly_full_alignment.bam
```
Comment (mahesh-panchal) [inline subworkflows/OrganelleFinder/modules/mapping.nf:23]:
```suggestion
    minimap2 -a assembly.mmi $reads_file | samtools sort --threads $task.cpus -o sorted_assembly_full_alignment.bam -
```
piping should help save some I/O. `samtools sort` is also multithreaded

**Review by mahesh-panchal — COMMENTED** [review]
A few more comments in addition to the last ones.

**Comment by mahesh-panchal** [comment]
Toy showing output only has to match pattern, and won't complain if the other is not there.
`main.nf`:
```groovy
nextflow.enable.dsl = 2

workflow {
    FOO(['mito','chlo'])
}

process FOO {
    input:
    each org

    output:
    path "*.txt"

    script:
    """
    touch file_${org}.txt
    """
}
```
`nextflow.config`:
```groovy
process {
    withName: 'FOO' {
        publishDir = [
            path: './results',
            pattern: "*_{mito,chlo}.txt"
        ]
    }
}
```
Each task makes a different file, but the pattern matches them both.

**Comment by mahesh-panchal** [comment]
And `{}` in `pattern:` doesn't work, but also doesn't error. It just silently matches nothing.

File: subworkflows/OrganelleFinder/nextflow.config [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+            name = 'slurm'
+        }
+        process {
+            scratch = '$SNIC_TMP'
+        }
+        singularity.enabled = true
+        singularity.envWhitelist = 'SNIC_TMP'
+    }
+
+process {
```
Comment (mahesh-panchal) [inline subworkflows/OrganelleFinder/nextflow.config:64]: This process configuration block is sitting inside the profiles block, so it won't be seen. You've configured nextflow to look for a profile named process here.

File: subworkflows/OrganelleFinder/.gitignore [hunk 9]
```diff
@@ -0,0 +1,6 @@
+.nextflow*
```
Comment (mahesh-panchal) [inline subworkflows/OrganelleFinder/.gitignore:1]: Remove the `.gitignore` file here.

File: subworkflows/OrganelleFinder/OrganelleFinder.nf [hunk 10]
```diff
@@ -0,0 +1,144 @@
+#! /usr/bin/env nextflow
+
+nextflow.enable.dsl=2
+
+/*
+Authored by Vendela, William and Viktor
```
Comment (mahesh-panchal) [inline subworkflows/OrganelleFinder/OrganelleFinder.nf:6]: Add your full names and github id's in brackets

File: subworkflows/OrganelleFinder/OrganelleFinder.nf [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+            blastfiles, params.mit_blast_evalue)
+        FILTER_MITOCHONDRIA(TBLASTN_MITOCHONDRIA.out.output_blast, params.mit_bitscore, params.mit_significant_gene_matches, params.mit_suspicious_gene_matches, params.mit_max_scaffold_length, params.mit_min_span_fraction, "mitochondria")
+        MAPPING(genome_assembly, reads_file)
+        DEPTH_FILTER_ANIMAL(MAPPING.out.depth_file, FILTER_MITOCHONDRIA.out.accessions, FILTER_MITOCHONDRIA.out.accessions_suspicious)
+        PLOTING(DEPTH_FILTER_ANIMAL.out.collect())
+        STATISTICS_MITOCHONDRIA(FILTER_MITOCHONDRIA.out.statistics, FILTER_MITOCHONDRIA.out.accessions, FILTER_MITOCHONDRIA.out.accessions_suspicious, "mitochondria")
+        EXTRACT_FINAL(genome_assembly,
+            FILTER_MITOCHONDRIA.out.accessions, "mitochondria")
+    emit:
+        blast_result = EXTRACT_FINAL.out[0]
```
Comment (mahesh-panchal) [inline subworkflows/OrganelleFinder/OrganelleFinder.nf:109]: Please add named outputs to the processes for readability.
```
EXTRACT_FINAL.out[0] -> EXTRACT_FINAL.out.blast_tsv
```

**Comment by mahesh-panchal** [comment]
Also when your final version is ready, change the pull request to Ready for review from a Draft.

File: subworkflows/OrganelleFinder/README.md [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+   * `chl_suspicious_gene_matches`: Threshold for number of unique matches for contig to be classified as suspicious (default: 2).
+   * `chl_max_contig_length`: Max bp length of contig to be classified as chloroplast (default: 400000)
+   * `chl_min_span_fraction`: Minimal match span fraction for contig to be classified as chloroplast (default: 0.8)
+
+
+Parameters to the workflow is provided by via a params-file that may look as follows:
+
+`animal_params.yml`:
+```
+// General parameters
```
Comment (LucileSol) [inline subworkflows/OrganelleFinder/README.md:44]:
```suggestion
# General parameters
```

File: subworkflows/OrganelleFinder/README.md [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+genome_assembly : ''
+reference_mitochondria : ''
+reads_file : ''
+input_type : 'animal'
+outdir: './results'
+
+// Mitochondrial parameters
```
Comment (LucileSol) [inline subworkflows/OrganelleFinder/README.md:51]:
```suggestion
# Mitochondrial parameters
```

File: subworkflows/OrganelleFinder/README.md (hunk 12 above)
Comment (LucileSol) [reply subworkflows/OrganelleFinder/README.md:44]: // is not commenting

**Comment by LucileSol** [comment]
I ran 
```nextflow run -profile singularity ../git/applied/pipelines-nextflow/subworkflows/OrganelleFinder/OrganelleFinder.nf -params-file 'plant_params.yml' ```

with this parameter file : 
```
# General parameters
genome_assembly : 'linum_all.fna'
reference_mitochondria : 'mitochondrion.1.1.genomic.fna'
reads_file : ''
input_type : 'plant'
outdir: './results'

# Mitochondrial parameters
mit_blast_evalue : '1e-6'
mit_bitscore : 100
mit_significant_gene_matches : 2
mit_suspicious_gene_matches : 1
mit_max_contig_length : 100000
mit_min_span_fraction : 0.8


reference_chloroplast: 'plastid.1.1.genomic.fna'
chl_blast_evalue : '1e-6'
chl_bitscore : 100
chl_significant_gene_matches : 2
chl_suspicious_gene_matches : 1
chl_max_contig_length : 100000
chl_min_span_fraction : 0.8
```
The pipeline ran but I have no results meaning that it did not find any chloroplast or mitochondria (and I am sure it contains mitochondria and chloroplast), the statistics files are also empty.

In my genome sequence I have : 

```
>scaffold_1--33206302_34585040+
>scaffold_2--33548729_34811913+
>scaffold_3--34545344+
>scaffold_4--34554814_34515665_34992229_35017581+
>scaffold_5--34585812+
>scaffold_6--34762752_34747320+
>scaffold_7--34823016_34429721+
>scaffold_8--34898971_34893665_35018255+
>scaffold_9--35017247+
>scaffold_10--35017267_35024205_34220946_34541214_35017265+
>scaffold_11--35027699+
>scaffold_12--35030837_34880665_35017409_35017471_34531555_34470340_34842134_35018041_35017541+
>LG10
>LG1
>LG2
>LG3
>LG4
>LG5
>LG6
>LG7
>LG8
>LG9
>3565841_3545355_3565803_3565631_3565707_3565589_3565747+,3565515_3565799_3565835_3549281_3565559_3565827_3565655_3565045_3563565_3565839+,3565815+,3565515_3565799_3565835_3549281_3565559_3565827_3565655_3565045_3563565_3565839-(circular)
>3565841_3545355_3565803_3565631_3565707_3565589_3565747+,3565515_3565799_3565835_3549281_3565559_3565827_3565655_3565045_3563565_3565839+,3565815-,3565515_3565799_3565835_3549281_3565559_3565827_3565655_3565045_3563565_3565839-(circular)
```
Where the first 12 contigs are mitochondria and the last two chloroplasts.

Is there some specific format that my contigs should follow? (uppercase/lowercase, number of nucleotides per line, naming ...)

**Review by mahesh-panchal — APPROVED** [review]
I'm approving, but I'd like to get a review from @LucileSol  to make sure the workflow is working.

File: subworkflows/OrganelleFinder/modules/makeblastdb.nf [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+              'quay.io/biocontainers/blast:2.12.0--pl5262h3289130_0' }"
+
+    label 'blast'
+
+    input:
+    path genome
+    val state
+
+    output:
+    path "*.fna*"
```
Comment (mahesh-panchal) [inline subworkflows/OrganelleFinder/modules/makeblastdb.nf:16]: This needs to be generalised to a fasta named either `*.fasta`, `*.fa` or `*.fna`.

File: subworkflows/OrganelleFinder/OrganelleFinder.nf [hunk 15]
```diff
@@ -0,0 +1,147 @@
...
+workflow PLANT_ORGANELLE_FINDER {
+
+    take:
+        genome_assembly
+        reference_mitochondria
+        reference_chloroplast
+        reads_file
+
+    main:
+        MAKEBLASTDB_MITOCHONDRIA(genome_assembly,genome_assembly.filter { it =~ /.p(hr|in|sq)$/ }.ifEmpty('DBFILES_ABSENT'))
```
Comment (mahesh-panchal) [inline subworkflows/OrganelleFinder/OrganelleFinder.nf:124]:
```suggestion
        MAKEBLASTDB_MITOCHONDRIA(genome_assembly,genome_assembly.filter { it =~ /.n(hr|in|sq)$/ }.ifEmpty('DBFILES_ABSENT'))
```
The filter removes protein database files, and this should be nuclear

**Comment by nylander** [comment]
Short comment on the status: current code did not run (@LucileSol ), and further testing is needed.

**Pull request closed without merging**